### PR TITLE
[DTensor] allow numel 1 tensor operand to be implicitly replicate DTensor

### PR DESCRIPTION
**Pull request by wz337**
Fixes #ISSUE_NUMBER

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @tianyu-l @wconstab @yf225 @chauhang @d4l3k @msaroufim @rohan-varma

### CODE REVIEW

File: torch/distributed/_tensor/dispatch.py [hunk 1]
```diff
@@ -297,7 +297,11 @@ def unwrap_to_op_info(
                 else:
                     mesh = arg.device_mesh
             elif isinstance(arg, torch.Tensor):
-                if arg.ndim == 0 or self._allow_implicit_replication:
+                if (
+                    arg.ndim == 0
+                    or arg.numel() == 1
```
Comment (wanchaol) [inline torch/distributed/_tensor/dispatch.py:302]: I think: we might want to throw some warnings for the `arg.numel() == 1` case, where if we found such a thing, in addition to creating a replicated DTensor, we also need to throw warning to the user to let user either change their code to a scalar tensor, or create it with DTensor when in the distributed context

Comment (wz337) [reply torch/distributed/_tensor/dispatch.py]: > I think: we might want to throw some warnings for the `arg.numel() == 1` case, where if we found such a thing, in addition to creating a replicated DTensor, we also need to throw warning to the user to let user either change their code to a scalar tensor, or create it with DTensor when in the distributed context

Sounds good! Updated the PR to include the warning.

**Comment by kwen2501** [comment]
If we put in special logic for `numel() == 1`, what about the case of `numel() < nranks`?

File: torch/distributed/_tensor/dispatch.py [hunk 2]
```diff
@@ -300,7 +301,16 @@ def unwrap_to_op_info(
...
+                if arg.numel() == 1 and arg.ndim == 1:
+                    warnings.warn(
+                        "Found a non-scalar tensor with numel=1 and ndim!=0, "
+                        "we are implicitly creating a replicated DTensor for it. "
+                        "However, please consider changing it to a scalar tensor "
+                        "or explicitly create a DTensor under distributed enviroment."
+                    )
+
+                # if the arg.numel() == 1, arg.ndim could be 0 or 1.
+                if arg.numel() == 1 or self._allow_implicit_replication:
```
Comment (wanchaol) [inline torch/distributed/_tensor/dispatch.py:313]: Oh I think this should be `arg.ndim == 0`? and we can make this an `elif`

File: test/distributed/_tensor/test_dtensor.py [hunk 3]
```diff
@@ -763,6 +763,25 @@ def test_implicit_replication(self):
...
+
+        # automatically turn tensor to DTensor replicate when ndim = 0 and numel = 1
+        ndim_0_tensor = torch.tensor(1, device=self.device_type)
+        self.assertEqual(
+            (sharded_dtensor + ndim_0_tensor).to_local(), local_tensor + ndim_0_tensor
+        )
+
+        # automatically turn tensor to DTensor replicate when ndim = 1 and numel = 1
+        numel_1_tensor = torch.tensor([1], device=self.device_type)
+        self.assertEqual(
```
Comment (wanchaol) [inline test/distributed/_tensor/test_dtensor.py:781]: can you add tests where if it's a "Scalar" Tensor, we don't throw warning? i.e. self.assertNoWarn

**Review by wanchaol — APPROVED** [review]
have two comments inlined, please address, stamp to unblock

**Comment by wz337** [comment]
> If we put in special logic for `numel() == 1`, what about the case of `numel() < nranks`?

For this case specifically, there is no sharding involved. This is gonna be a replicated DTensor. So it's ok that numel() = 1 < nranks. 

For all the other cases, numel() < nranks case is handled in here:
https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/placement_types.py#L59-L61

File: torch/distributed/_tensor/dispatch.py (hunk 2 above)
Comment (wz337) [reply torch/distributed/_tensor/dispatch.py]: We are turning the tensor with numel() = 1 to a replicated DTensor when arg.ndim <= 1. Hence, I didn't use an elif since we are appending the same args_schema for both cases. Therefore, I add an additional check before this to log the warnings when arg.ndim = 1.

**Comment by kwen2501** [comment]
> For all the other cases, numel() < nranks case is handled in here: https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/placement_types.py#L59-L61

Not sure how the pointed assert handles the case -- it only checks `sharder.dim` against `tensor.ndim`.
